src/Carbon/CarbonPeriod.php

```diff
@@ -0,0 +1,1623 @@
+     *
+     * @return void
+     */
+    protected function initializeCurrentDate()
+    {
+        $this->current = $this->startDate->copy();
+
+        if (static::intervalHasTime($this->dateInterval)) {
+            $this->timezone = $this->current->getTimezone();
+
+            $this->current->setTimezone('UTC');
+        } else {
+            $this->timezone = null;
+        }
+    }
+
+    /**
+     * Keep incrementing the current date until a valid date is found or the iteration is ended.
+     *
+     * @throws \RuntimeException
+     *
+     * @return void
+     */
+    protected function incrementCurrentDateUntilValid()
+    {
+        $attempts = 0;
+
+        do {
+            $this->current->add($this->dateInterval);
+
+            $this->validationResult = null;
+
+            if (++$attempts > static::NEXT_MAX_ATTEMPTS) {
+                throw new RuntimeException('Could not find next valid date.');
+            }
+        } while ($this->validateCurrentDate() === false);
+    }
+
+    /**
+     * Set current date from the iteration results and return true on success.
+     *
+     * @param int $key
+     *
+     * @return bool
+     */
+    protected function setCurrentDateFromCache($key)
+    {
+        if ($this->parametersChanged || $this->isCacheDisabled()) {
+            return false;
+        }
+
+        if ($key < count($this->iterationResults)) {
+            $this->validationResult = true;
+
+            $this->key = $key;
+            $this->current = $this->iterationResults[$key]->copy();
+
+            return true;
+        }
+
+        if ($this->iterationCompleted) {
+            $this->validationResult = static::END_ITERATION;
+
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Format the date period as ISO 8601.
+     *
+     * @return string
+     */
+    public function toIso8601String()
+    {
+        $parts = array();
+
+        if ($this->recurrences !== null) {
+            $parts[] = 'R'.$this->recurrences;
+        }
+
+        if ($this->startDate !== null) {
+            $parts[] = $this->startDate->toIso8601String();
+        }
+
+        $parts[] = $this->dateInterval->spec();
+
+        if ($this->endDate !== null) {
+            $parts[] = $this->endDate->toIso8601String();
+        }
+
+        return implode('/', $parts);
+    }
+
+    /**
+     * Convert the date period into a string.
+     *
+     * @return string
+     */
+    public function toString()
+    {
+        $translator = Carbon::getTranslator();
+
+        $parts = array();
+
+        if ($this->startDate && !$this->startDate->isStartOfDay() || $this->endDate && !$this->endDate->isStartOfDay()) {
+            $format = 'Y-m-d H:i:s';
+        } else {
+            $format = 'Y-m-d';
+        }
+
+        if ($this->recurrences !== null) {
+            $parts[] = $translator->transChoice('period_recurrences', $this->recurrences, array(':count' => $this->recurrences));
```

Comment:
It's ambiguous, recurrences are occurrences + 1. Can be kept as is for now, but In fact, it would be something more like:
occurrences = recurrences + (exclude_start ? 1 : 0)
- occurrences times
- exclude_start ? from : after
- exclude_end ? to : until

But it can make things more confused.

Reply:
It for sure confused me :) If you want to change it I'd advocate keeping the behavior and renaming `recurrences` to `limit` or something like that.

Reply:
What I say is: start monday 3 recurrences of 1 day will give `3 times every 1 day from monday` when the CarbonPeriod iterate 4 times. That's why the `"times"` word may not be the good one or else the way to count.

Reply:
That reflects the actual behavior - 3 recurrences in CarbonPeriod will return 3 results. What confused me was DatePeriod which would give me 3 results for 3 recurrences with EXCLUDE_START but 4 results without EXCLUDE_START.

Thus I think about renaming the feature. Particularly that recurrences can be mixed with end date and other filters (for example 01-10 → 01-11 despite 5 recurrences gives 2 results). So "limit" would be more accurate.

Could be made also into separate features to keep the standard behavior of recurrences, but provide limit functionality.